### ISSUE TITLE
minimum write DataRate 

### ISSUE THREAD

**Issue by gregw** [eclipse/jetty.project#1973]
There is currently an optional minimum read data rate, but no minimum write data rate.
Consideration of the blocking timeout from #1958 has indicated that a data rate is perhaps a better mechanism for slow client protection

**Comment by gregw** [comment]
It may also be worth while considering moving data rate mechanism to input/output interceptors. Then multiple instances could be added with warnings/errors at different thresholds

**Comment by joakime** [comment]
The input/output interceptor approach would also allow for custom handlers/filters to decide which requests to add those to, possibly even with different configurations (eg: time critical responses vs long uploads vs slow database queries).

**Comment by joakime** [comment]
@gregw one downside to input/output interceptor approach is that these rates would not apply to request headers. right?

**Comment by gregw** [comment]
@joakime correct - although currently the input data rate is also only in HttpInput and thus also does not apply to headers.         Note also that the blocking timeout doesn't apply to headers either.

So a data rate lower down might be worthwhile.  Perhaps a datarate listener in the endpoint?

**Comment by sbordet** [comment]
@gregw not sure I want data rate for headers, as they will be read non-blocking and we have not called the application yet, so I don't think it's a risk like it would be for blocking content read performed by the application.

**Comment by gregw** [comment]
@sbordet good point.  Data rate minimum is only there to protect blocking applications from wasting threads.

**Comment by joakime** [comment]
minimum dataRate is not only to protect blocking application usage.
minimum dataRate also protects (in a general sense) from malicious clients that are (intentionally?) being slow.

**Comment by gregw** [comment]
@joakime exactly.  But because we handle headers asynchronously a malicious client can't take many resources by sending a header slowly. In fact the DosFilter uses async waiting to hold up a malicious client.

So a deliberately slow client only really hurts the server if we allocate a thread for blocking IO - hence we need the data rate to protect an application reading the content, but not the headers.

**Comment by gregw** [comment]
The implementation of the response data rate has to be within the WriteFlusher, as this is the only place that will see regular iterations while write very large content.  Work in progress is in PR #1988
